src/_macosx.m

```diff
@@ -2066,29 +2097,87 @@ - (BOOL)acceptsFirstResponder
     return YES;
 }
 
-/* This is all wrong. Address of pointer is being passed instead of pointer, keynames don't
-   match up with what the front-end and does the front-end even handle modifier keys by themselves?
-
-- (void)flagsChanged:(NSEvent*)event
-{
-    const char *s = NULL;
-    if (([event modifierFlags] & NSControlKeyMask) == NSControlKeyMask)
-        s = "control";
-    else if (([event modifierFlags] & NSShiftKeyMask) == NSShiftKeyMask)
-        s = "shift";
-    else if (([event modifierFlags] & NSAlternateKeyMask) == NSAlternateKeyMask)
-        s = "alt";
-    else return;
-    PyGILState_STATE gstate = PyGILState_Ensure();
-    PyObject* result = PyObject_CallMethod(canvas, "key_press_event", "s", &s);
-    if(result)
-        Py_DECREF(result);
-    else
-        PyErr_Print();
-
-    PyGILState_Release(gstate);
+// flagsChanged gets called whenever a  modifier key is pressed OR released
+// so we need to handle both cases here
+- (void)flagsChanged:(NSEvent *)event
+{
+    bool keypress = false;
+    bool keyrelease = false;
+    if (([event modifierFlags] & NSEventModifierFlagCommand) && !lastCommand) {
+        // Command pressed
+        lastCommand = true;
+        keyChangeCommand = true;
+        keypress = true;
+    }
+    else if (!([event modifierFlags] & NSEventModifierFlagCommand) && lastCommand) {
+        // Command released
+        lastCommand = false;
+        keyChangeCommand = true;
+        keyrelease = true;
+    }
```

Comment:
Not a huge deal, but I guess each of these pairs can be grouped to
```objc
bool isPress = true;  // false for release
if ([event modifierFlags] & NSEventModifierFlagCommand) {
    lastCommand = !lastCommand
    keyChangeCommand = true
    isPress = lastCommand;
} else if (...)
    ...
} else {  // Unknown flag
    return;
}
```

Reply:
`NSEventModifierFlagCommand` will only be true when it is held down, so we would miss the release with that implementation I think... We could probably squash the two cases into a single if block with your suggestion, but maybe I'm missing a more elegant way to keep track of the push/release here:

```objc
    if ((([event modifierFlags] & NSEventModifierFlagCommand) && !lastCommand) ||
        (!([event modifierFlags] & NSEventModifierFlagCommand) && lastCommand)) {
        // Command pressed or released
        lastCommand = !lastCommand;
        keyChangeCommand = true;
        keypress = lastCommand;
    } ...
```

Reply:
Ah, sorry, I missed the inversion.  Perhaps a good place for a `^` (xor)?

Reply:
:+1: Yep, a good place for it, thanks.

Reply:
LGTM, although probably another mac user should test :-)